### DISCUSSION TITLE
Is there a supported way to display the output of dagstermill notebooks in kubernetes?

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#17659] [Q&A] [answered]
Is there a supported way to display the output of dagstermill notebooks in kubernetes?

The built-in `ConfigurableLocalOutputNotebookIOManager` works great locally, but results in an internal server error when running in kubernetes (it looks as if the notebook gets written to the ephemeral disk of a k8s Job)

Other, non-dagstermill-specific IO managers like the s3 pickle IO manager don’t display the “view notebook” link in the dagster UI (it looks as if the IO manager needs to return some notebook specific metadata for this)

I started writing a custom IO manager to do this, but I’m having trouble figuring out exactly what interface it needs to satisfy for the notebook display to work

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1698955448294449?thread_ts=1698955448.294449&cid=C01U954MEER).*

**Comment by yuhan** [comment, marked as answer]
You may [record metadata from a custom I/O manager](https://docs.dagster.io/concepts/io-management/io-managers#recording-metadata-from-an-io-manager), where the metadata value can be `NotebookMetadataValue` ([apidoc](https://docs.dagster.io/_apidocs/ops#dagster.MetadataValue.notebook)):
```python
class MyIOManager(ConfigurableIOManager):
    def handle_output(self, context, obj):
        ...
        context.add_output_metadata(
            metadata={
                "notebook_path": MetadataValue.notebook("path/to/notebook.ipynb"),
            }
        )

    def load_input(self, context):
        ...

```

**Comment by kztwigex** [comment]
Hi,

Any success on this?
I forked existing [s3 io-manager ](https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py)and used it as base for notebook output manager. Adding metadata helped to show "View notebook" link.

But it will NOT actually download/open notebook from S3. Dagster's machinery is not capable (has no functionality) to load notebooks from S3? It only can load notebooks that are saved locally in k8s pod.

At least it is so in code:

https://github.com/dagster-io/dagster/blob/2582697b04f92a7729cbe58fa1dfcfbac56a074f/python_modules/dagster/dagster/_grpc/impl.py#L601

```

def get_notebook_data(notebook_path):
    check.str_param(notebook_path, "notebook_path")

    if not notebook_path.endswith(".ipynb"):
        raise Exception(
            "unexpected file extension for notebooks. Please provide a path that ends with"
            " '.ipynb'."
        )

    with open(os.path.abspath(notebook_path), "rb") as f:
        content = f.read()
        return content

```

Does anyone knows how to show notebooks that are saved on S3?

**Comment by danielgafni** [comment]
I think you guys would benefit from: 

```python
from upath import UPath

data = UPath("s3://path/to/tonebook.ipynb").read_bytes()
```

Would work with other storages (like GCS) too. 

You can also subclass `dagster.UPathIOManager` and implement `dump_to_path` and `load_from_path`.
